src/ray/core_worker/transport/dependency_resolver.cc

```diff
@@ -70,50 +80,74 @@ void InlineDependencies(
+void LocalDependencyResolver::ResolveDependencies(
+    TaskSpecification &task, std::function<void(Status)> on_complete) {
   absl::flat_hash_map<ObjectID, std::shared_ptr<RayObject>> local_dependencies;
+  std::vector<ActorID> actor_dependences;
   for (size_t i = 0; i < task.NumArgs(); i++) {
     if (task.ArgByRef(i)) {
       local_dependencies.emplace(task.ArgId(i), nullptr);
     }
+    for (const auto &in : task.ArgInlinedRefs(i)) {
+      auto object_id = ObjectID::FromBinary(in.object_id());
+      if (ObjectID::IsActorID(object_id)) {
+        auto actor_id = ObjectID::ToActorID(object_id);
+        if (actor_creator_.IsActorInRegistering(actor_id)) {
+          actor_dependences.emplace_back(ObjectID::ToActorID(object_id));
+        }
+      }
+    }
   }
-  if (local_dependencies.empty()) {
-    on_complete();
+  if (local_dependencies.empty() && actor_dependences.empty()) {
+    on_complete(Status::OK());
     return;
   }
 
   // This is deleted when the last dependency fetch callback finishes.
-  std::shared_ptr<TaskState> state =
-      std::make_shared<TaskState>(task, std::move(local_dependencies));
+  std::shared_ptr<TaskState> state = std::make_shared<TaskState>(
+      task, std::move(local_dependencies), std::move(actor_dependences));
   num_pending_ += 1;
 
   for (const auto &it : state->local_dependencies) {
     const ObjectID &obj_id = it.first;
-    in_memory_store_->GetAsync(obj_id, [this, state, obj_id,
-                                        on_complete](std::shared_ptr<RayObject> obj) {
+    in_memory_store_.GetAsync(obj_id, [this, state, obj_id,
+                                       on_complete](std::shared_ptr<RayObject> obj) {
       RAY_CHECK(obj != nullptr);
       bool complete = false;
       std::vector<ObjectID> inlined_dependency_ids;
       std::vector<ObjectID> contained_ids;
       {
         absl::MutexLock lock(&mu_);
         state->local_dependencies[obj_id] = std::move(obj);
-        if (--state->dependencies_remaining == 0) {
+        if (--state->obj_dependencies_remaining == 0) {
           InlineDependencies(state->local_dependencies, state->task,
                              &inlined_dependency_ids, &contained_ids);
-          complete = true;
-          num_pending_ -= 1;
+          if (state->actor_dependencies_remaining == 0) {
+            complete = true;
+            num_pending_ -= 1;
+          }
         }
       }
       if (inlined_dependency_ids.size() > 0) {
-        task_finisher_->OnTaskDependenciesInlined(inlined_dependency_ids, contained_ids);
+        task_finisher_.OnTaskDependenciesInlined(inlined_dependency_ids, contained_ids);
       }
       if (complete) {
-        on_complete();
+        on_complete(state->status);
       }
     });
   }
+
+  for (const auto &actor_id : state->actor_dependencies) {
+    actor_creator_.AsyncWaitForActorRegisterFinish(
+        actor_id, [state, on_complete](Status status) {
+          if (!status.ok()) {
```

Comment:
Is this thread-safe?

Reply:
(do we need mutexLock(mu_))?

Reply:
@ericl I think you are asking a really good question and I found out that I don't have a clear answer unless I'm checking the code e2e.

I think the issue here is the poorly defined threading model. I probably will re-investigate the fiber thing mentioned previously after this.

For this question:
CoreWorker::SubmitActorTask -> direct_actor_submitter_->SubmitTask(io pool) -> ResolveDependency 
CoreWorker::SubmitTask -> direct_task_submitter_->SubmitTask(io pool) -> ResolveDependency
CoreWorker::CreateActor -> direct_task_submitter_->SubmitTask -> ResolveDependency

So create actors doesn't work. I'll update create actor to post tasks into io pool like SubmitActorTask/SubmitTask is doing.

Reply:
Ok. If we're relying on tasks running in a particular thread for safety, we could add asserts that the current thread id is a particular one, this would also help the readability of the code.

Reply:
I think that will help identify whether things are going wrong, but not very useful for the organization of the code and lead to poor readability. 

We should have a clear boundary between modules running in different threads and adopt a way that prevents developer from doing things incorrectly.